Introducting GBPTree, initial step for native indexing

burqen opened the pull request:
**WHAT?**
This is the first step in building a generation-aware B+tree (GB+Tree) implementation directly atop a PageCache with no caching in between. Additionally internal and leaf nodes on same level are linked both left and right with sibling pointers, this to provide correct reading when concurrently modifying the tree.

**WHY?**
Initial prototyping has shown great performance improvement on label scan store writes compared to lucene.

A GB+Tree could potentially be used for a lot of different types of indexes, path index for example.

**DETAILS ON IMPLEMENTATION**
*Non blocking*
Tree is modified by a single writer, multiple concurrent readers are allowed. Both reading and writing is non blocking.

*Garbage*
Implementation aim to make reading and writing garbage free.

*Build order*
The 'index' module is built directly before 'kernel' module, see community package pom.

**GENERATION**
Every node and child/sibling pointer has a generation indicating when it first evolved (was created). During recovery, pointers and nodes can have been flushed to disk or lost from memory. To determine if a pointer or node can be 'trusted' to be stable on disk there is a global STABLE generation of the tree. This generation and every generation before it has been checkpointed (flushed to disk) and can therefore be trusted, even after a crash.

The generation currently under evolution is considered to be UNSTABLE. Data belonging to or pointing to UNSTABLE generation can not be trusted after a crash and must be overwritten during recovery. Note that this is not implemented yet. After checkpoint the STABLE generation will be set to the UNSTABLE generation and UNSTABLE generation will be incremented.

*GenSafePointer, GSP*
By guarding the pointer fields, generation and address, with a checksum we can determine if a pointer is BROKEN (wrong checksum). A pointer can be broken if the page is flushed while writing the pointer and then crash before that page is flushed again.

A pointer with generation and checksum is called a GenSafePointer, GSP.

*GenSafePointerPair, GSPP*
We always want to keep a stable version of every pointer. Therefore all pointers are in fact two GSPs, a GSPP. So when we read or write a child pointer for example, we look at the pair for pointers as a whole and determine which one to read from or write to, look at GenSafePointerPair class.

**LabelScanStore**
There is an implementation of a label scan store built on top of this that will be introduced in a separate PR.

**Note**
This is minimal feature set and among the things missing (but in pipeline) are:
- Implement recovery
- Implement freelist
- Rebalance and merge when removing from tree

tinwelint committed:
Moves Exceptions, Randoms and RandomRule to neo4j-common

tinwelint committed:
Introducting GBPTree, initial step for native indexing

**WHAT?**
This is the first step in building a generation-aware B+tree (GB+Tree) implementation directly atop a PageCache with no caching in between. Additionally internal and leaf nodes on same level are linked both left and right with sibling pointers, this to provide correct reading when concurrently modifying the tree.

**WHY?**
Initial prototyping has shown great performance improvement on label scan store writes compared to lucene.

A GB+Tree could potentially be used for a lot of different types of indexes, path index for example.

**DETAILS ON IMPLEMENTATION**
*Non blocking*
Tree is modified by a single writer, multiple concurrent readers are allowed. Both reading and writing is non blocking.

*Garbage*
Implementation aim to make reading and writing garbage free.

*Build order*
The 'index' module is built directly before 'kernel' module, see community package pom.

**GENERATION**
Every node and child/sibling pointer has a generation indicating when it first evolved (was created). During recovery, pointers and nodes can have been flushed to disk or lost from memory. To determine if a pointer or node can be 'trusted' to be stable on disk there is a global STABLE generation of the tree. This generation and every generation before it has been checkpointed (flushed to disk) and can therefore be trusted, even after a crash.

The generation currently under evolution is considered to be UNSTABLE. Data belonging to or pointing to UNSTABLE generation can not be trusted after a crash and must be overwritten during recovery. Note that this is not implemented yet. After checkpoint the STABLE generation will be set to the UNSTABLE generation and UNSTABLE generation will be incremented.

*GenSafePointer, GSP*
By guarding the pointer fields, generation and address, with a checksum we can determine if a pointer is BROKEN (wrong checksum). A pointer can be broken if the page is flushed while writing the pointer and then crash before that page is flushed again.

A pointer with generation and checksum is called a GenSafePointer, GSP.

*GenSafePointerPair, GSPP*
We always want to keep a stable version of every pointer. Therefore all pointers are in fact two GSPs, a GSPP. So when we read or write a child pointer for example, we look at the pair for pointers as a whole and determine which one to read from or write to, look at GenSafePointerPair class.

**LabelScanStore**
There is an implementation of a label scan store built on top of this that will be introduced in a separate PR.

**Note**
This is minimal feature set and among the things missing (but in pipeline) are:
- Implement recovery
- Implement freelist
- Rebalance and merge when removing from tree

chrisvest commented on community/index/src/main/java/org/neo4j/index/Index.java:
```diff
@@ -0,0 +1,65 @@
+     * @return {@link Modifier} able to modify the index.
+     * @throws IOException on error accessing the index.
+     */
+    Modifier<KEY,VALUE> modifier( Modifier.Options options ) throws IOException;
```
I think `IndexWriter` or `IndexMutator` would be better names. `Modifier` sounds like something that offsets the result of a computation.

chrisvest commented on community/index/src/main/java/org/neo4j/index/Modifier.java:
```diff
@@ -0,0 +1,115 @@
+     * @param amender {@link ValueAmender} to consult if key already exists.
+     * @throws IOException on index access error.
+     */
+    void insert( KEY key, VALUE value, ValueAmender<VALUE> amender ) throws IOException;
```
This is equivalent to the `java.util.Map#merge` method. Perhaps we can borrow nomenclature from there? Also, take a look at the types they are using.

chrisvest commented on community/index/src/main/java/org/neo4j/index/Index.java:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @throws IOException on error flushing to storage.
+     */
+    void flush() throws IOException;
```
Is this really going to flush things, or just update the generations?

burqen replied:
It should also flush, but that part is not implemented yet.
A call to this flush will be the last step when checkpointing the index.

chrisvest replied:
Okay, in that case you should make it take, and pass an `IOLimiter` to the `PagedFile#flushAndForce` method.

burqen replied:
Good point! We will add this when we implement checkpointing.

chrisvest commented on community/index/src/main/java/org/neo4j/index/gbptree/GenSafePointer.java:
```diff
@@ -0,0 +1,132 @@
+ * All interaction is made using a {@link PageCursor}. These methods are about a single GSP,
+ * whereas the normal use case of a GSP is in pairs ({@link GenSafePointerPair GSPP}).
+ * <p>
+ * A GSP consists of [generation,pointer,checksum] where checksum is updated
```
Truncated sentence?

chrisvest commented on community/index/src/test/java/org/neo4j/index/gbptree/GBPTreeTest.java:
```diff
@@ -0,0 +1,624 @@
+    public GBPTree<MutableLong,MutableLong> createIndex( int pageSize )
+            throws IOException
+    {
+        pageCache = new MuninnPageCache( swapperFactory(), 10_000, pageSize, NULL );
```
Why don't you use the page cache rule?

tinwelint replied:
Probably because it lives in kernel, but it could be moved of course.

chrisvest reviewed (commented):
Had a quick glance through, and had a few comments on obvious stuff I happened upon.

Commit:
Change name for Modifier to IndexWriter

Also updated outdated variable names for (now) IndexWriter and InternalTreeLogic

Commit:
Split insert in IndexWriter into merge and put

Taking inspiration from Map#merge.
Removed ValueAmender (now ValueMerger) that could insert non unique keys, because this is not supported by the tree.

Commit:
Javadoc fix for GenSafePointer

MishaDemianenko reviewed (changes requested):
First part of review couple of comments so far

and all my comments gone somewhere...so i will start all over again! feels_like_a_monday! doooh

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/Index.java:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @throws IOException on error flushing to storage.
+     */
+    void flush() throws IOException;
```
It really strange to have flush method ad part of index interface. Since index already have a writer that ` able to modify the index, i.e. insert and remove keys/values` why it's not a writer responsibility then? also from a usage patter we do that only while closing index so why do not use close instead then?

burqen replied:
Flush will later be part of the checkpointing routine which will be
- flush index
- lock index (blocking writers)
- flush index (to make sure we flush data that was written during the previous flush)
- update index state (increment generation)
- flush index
- unlock index

It will of course also flush the page.

Does this make more sense?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/IndexWriter.java:
```diff
@@ -0,0 +1,111 @@
+        class Defaults implements Options
+        {
+            @Override
+            public float splitLeftChildSize()
```
it more sounds like a split factor and not a split size because it hard to think about a size of 0.5f and because we already referring to it as a factor in the java docs

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/IndexWriter.java:
```diff
@@ -0,0 +1,111 @@
+         * Keys (and its values/children) including the split key will go to the right tree node,
+         * everything before it goes into the left.
+         *
+         * @return a factor between 0..1 where 0 means far to the left, 1 means far to the right and
```
It gives a good explanation where we will put value (left or right) depending from supplied value but it does not describe macro effect of that: what impact that will give to the whole index

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/ValueMerger.java:
```diff
@@ -0,0 +1,39 @@
+     * @param withValue new value
+     * @return {@code value}, now merged with {@code withValue}, or {@code null} if no merge was done.
+     */
+    VALUE merge( VALUE value, VALUE withValue );
```
Can we rename parameters to oldValue and newValue instead?

tinwelint replied:
Sure, that makes sense. Or even, to be even more specific to the use case, `existingValue` and `newValue`?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/ValueMerger.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @param value existing value
+     * @param withValue new value
+     * @return {@code value}, now merged with {@code withValue}, or {@code null} if no merge was done.
```
why do we need to return null in case if nothing has been change? to check for null afterwards? can we simply return old value as result in case we want to preserve it?

burqen replied:
We took inspiration from `Map#merge` where a `null` from provided `BiFunction` means remove entry from map. We can not do that because we have a different path for remove. So we want to explicitly define the expected behaviour for when `ValueMerger` returns null.

We made it like this to avoid confusion, so it actually creates more confusion, it should be changed. Does it?

Note: Returning the old value will result in a write with no effective change of the actual value. We could add a check to avoid writing in that case.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/ValueMergers.java:
```diff
@@ -0,0 +1,66 @@
+public class ValueMergers
+{
+    @SuppressWarnings( "rawtypes" )
+    private static final ValueMerger OVERWRITE = new ValueMerger()
```
Lambda N1

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/ValueMergers.java:
```diff
@@ -0,0 +1,66 @@
+    };
+
+    @SuppressWarnings( "rawtypes" )
+    private static final ValueMerger KEEP_EXISTING = new ValueMerger()
```
Lambda N2

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/IndexWriter.java:
```diff
@@ -0,0 +1,111 @@
+ * @param <KEY> type of keys
+ * @param <VALUE> type of values
+ */
+public interface IndexWriter<KEY,VALUE> extends Closeable
```
Somewhere in indexWriter on in the index javadoc we need to mention what it the visibility of updated/inserted values? when they will be visible to the reader? are updates atomic?

tinwelint replied:
That'll most likely be implementation-specific so I'd say not suitable as interface javadoc. We can however elaborate on that in the `GBPTree` javadocs.

MishaDemianenko replied:
sure :)

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/ByteArrayPageCursor.java:
```diff
@@ -0,0 +1,278 @@
+import org.neo4j.io.pagecache.PageCursor;
+
+/**
+ * This class is bridging something which would otherwise make {@link InternalTreeLogic} code slightly more
```
I find it super hard to follow this java doc it gives a lot of historical insight and describe challenges that we faced somewhere else, but almost nothing about this class itself

tinwelint replied:
OK, thought it was so straight forward that it didn't need that kind of docs. We can add some about what it is and more notably what it can't do (w/ respect to PageCursor inheritance).

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/Layout.java:
```diff
@@ -0,0 +1,167 @@
+     * @param into instance to write minimum key into.
+     * @return the provided {@code into} instance for convenience.
+     */
+    KEY minKey( KEY into );
```
For minKey and maxKey why do we need to supply into param? can we simply return new value with min and max key?

burqen replied:
To avoid object creation. This is the general pattern used in layout. See `readKey` and `readValue`.

MishaDemianenko replied:
well except before invoking those now we always create new key for them :)

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/Layout.java:
```diff
@@ -0,0 +1,167 @@
+     * @param key key (left unchanged as part of this call) to copy contents from.
+     * @param into key (changed as part of this call) to copy contents into.
+     */
+    void copyKey( KEY key, KEY into );
```
Why it's index layout responsibility to copy keys?

burqen replied:
Because he is the only one who knows about how a key looks.

MishaDemianenko replied:
waaait, should key itself knew that?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/Layout.java:
```diff
@@ -0,0 +1,167 @@
+            upperInt |= (byteValue & 0xFF);
+        }
+
+        return upperInt << Integer.SIZE | (checksum & 0xFFFFFFFF);
```
Why do we need 0xFFFFFFFF here?

tinwelint replied:
To avoid negative values writing a lot of 1's into the checksum. Collision rate is orders of magnitude bigger on random values when not having this mask.

Commit:
Use lambdas in ValueMergers

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+                        minorVersion = metaCursor.getInt();
+                        layout.readMetaData( metaCursor );
+                    }
+                    while ( metaCursor.shouldRetry() );
```
`metaCursor.checkAndClearCursorException();`  should be added here ?

tinwelint replied:
Not until we, if at all, start setting cursor exceptions in our write execution path. To my knowledge the cursor exception is completely user controlled and serves as a way for application code to transfer an exception through a "shouldRetry" barrier.

Commit:
Clarified how keys are divided when splitting

MishaDemianenko commented on community/index/src/test/java/org/neo4j/index/gbptree/SimpleLongLayout.java:
```diff
@@ -0,0 +1,171 @@
+
+class SimpleLongLayout implements Layout<MutableLong,MutableLong>
+{
+    private static final Charset UTF_8 = Charset.forName( "UTF-8" );
```
this constant is useless `StandardCharsets.UTF_8` should be used instead

tinwelint replied:
Not useless, merely sub-optimal then :) we can certainly change that.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+     *
+     * @param pageCache {@link PageCache} to use to map index file
+     * @param indexFile {@link File} containing the actual index
+     * @param tentativePageSize page size, i.e. tree node size. Must be less than or equal to that of the page cache.
```
now i can see that, so please ignore this comment

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+            }
+            catch ( Throwable t )
+            {
+                try
```
IOUtils.closeAll

tinwelint replied:
Don't think that applies. We're closing only one thing and the exception we're potentially adding a suppressed exception to has already been thrown (and caught) and cannot be injected into closeAll, can it?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+                throw t;
+            }
+        }
+        catch ( NoSuchFileException e )
```
instead of catching exception shouldn't we check file existence instead to perform index creation?

tinwelint replied:
Letting the page cache throwing this is actually better because of potentially different hardware configuration the page cache may have. A file system instance has a more limited view of the world, compared to the page cache.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+                metaCursor.putLong( layout.identifier() );
+                metaCursor.putInt( layout.majorVersion() );
+                metaCursor.putInt( layout.minorVersion() );
+                layout.writeMetaData( metaCursor );
```
metaCursor retry, reset offset, check for errors stuff?

tinwelint replied:
Afaik not required when having a write lock on the cursor

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+        catch ( NoSuchFileException e )
+        {
+            pageSize = pageSizeForCreation == 0 ? pageCache.pageSize() : pageSizeForCreation;
+            if ( pageSize > pageCache.pageSize() )
```
exception handling should be extracted

tinwelint replied:
Are you thinking of the two (almost identical) exceptions around page size? Those two messages are phrased slightly different, one relevant for creation and the other for open. I think that's fine to have explicit and "duplicated".

Commit:
Better argument names for ValueMerger

Commit:
Use StandardCharset.UTF_8 instead Charset.forName( "UTF_8" )

tinwelint committed:
Improved javadoc of ByteArrayPageCursor

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+        {
+            do
+            {
+                isInternal = bTreeNode.isInternal( cursor );
```
in case of shouldRetry cursor offset will not be reseted

burqen replied:
cursor will do it himself in case of should retry

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+                isInternal = bTreeNode.isInternal( cursor );
+                // Find the left-most key within from-range
+                keyCount = bTreeNode.keyCount( cursor );
+                pos = 0;
```
it looks like we do not need this assignemnt

Commit:
Added part about visibility level in GBPTree javadoc

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/KeySearch.java:
```diff
@@ -0,0 +1,125 @@
+
+    private static int searchResult( int pos, boolean hit )
+    {
+        return (pos & 0x7FFFFFFF) | ((hit ? 1 : 0) << 31);
```
we should have a constants for all maks

Commit:
Remove minKey and maxKey from layout interface

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+    @Override
+    public RawCursor<Hit<KEY,VALUE>,IOException> seek( KEY fromInclusive, KEY toExclusive ) throws IOException
+    {
+        long localRootId = rootId;
```
i do not think we need localRootId variable in a for it's used now, it's only brings additional confusion why it's need to be reassigned.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+
+                PointerChecking.checkChildPointer( childId );
+
+                if ( !cursor.next( childId ) )
```
what if cursor.next faces io problems that should be retried?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+    // Utility method
+    public boolean consistencyCheck() throws IOException
+    {
+        try ( PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_READ_LOCK ) )
```
should retry?

burqen replied:
Yes, definitely. Suggest postponing this work for a different PR, as we need to refactor entire ConsistencyChecker to not be recursive. Card created.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+    @Override
+    public void flush() throws IOException
+    {
+        try ( PageCursor cursor = openMetaPageCursor( pagedFile ) )
```
should retry?

burqen replied:
Cursor has write lock, should retry not needed.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+    @Override
+    public IndexWriter<KEY,VALUE> writer( IndexWriter.Options options ) throws IOException
+    {
+        SingleIndexWriter result = this.writer.getAndSet( null );
```
mostly just an idea and own preference: what if instead of atomic reference for writer we will have ring buffer of work that writer will pick up instead?

burqen replied:
This could be a nice thing for the future if interaction with neo changes and we can't rely on WorkSync. For now, this implementation is bare minimum of what is needed.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -0,0 +1,481 @@
+        return result.take( rootId, options );
+    }
+
+    class SingleIndexWriter implements IndexWriter<KEY,VALUE>
```
writer does not have at all retry attempts or error handling for cursors problems

burqen replied:
Does not need to call `shouldRetry` because all page access is guarded by write locks. `InternalTreeLogic` does not set any cursor exceptions that we need to check. 
`checkAndClearBoundsFlag` should be moved to end of `SingleIndexWriter#merge`.

burqen replied:
Will fix `shouldRetry` for remove. Good catch!

Commit:
Cleanup of GBPTree

Commit:
Cleanup in KeySearch

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/RightmostInChain.java:
```diff
@@ -0,0 +1,77 @@
+        long rightSibling = node.rightSibling( cursor, stableGeneration, unstableGeneration );
+
+        // Assert we have reached expected node and that we agree about being siblings
+        assert leftSibling == currentRightmost :
```
should all those assertion be assertion exception instead of assertion errors?

MishaDemianenko replied:
thats applicable to whole consistency checker

burqen replied:
Yes, this should be fixed. Right now only used in tests and we are not sure how we want it to look, therefore we suggest postponing changing this.

Commit:
Better error handling when reading out of bounds

And removed unnecessary "go back to root retry"

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GenSafePointer.java:
```diff
@@ -0,0 +1,132 @@
+
+    public static long readGeneration( PageCursor cursor )
+    {
+        return cursor.getInt() & 0xFFFFFFFFL;
```
max generation mask?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GenSafePointerPair.java:
```diff
@@ -0,0 +1,485 @@
+    static final int STATE_SHIFT_B = 54;
+
+    // Aggregations and masks
+    static final long SUCCESS_WRITE_TO_B = SUCCESS | WRITE | WRITE_TO_B;
```
you can completely avoid | with SUCCESS since that is 0

MishaDemianenko replied:
same for GEN_EQUAL

burqen replied:
Yes, but we wanted to keep it here for code clarity

tinwelint replied:
Yeah, I'm pretty sure some level of the compiler or JIT will remove those anyway. It looks great when reading the code.

Commit:
Move checkAndClearBoundsFlag after IndexWriter has finished all writes

MishaDemianenko commented:
@burqen @tinwelint i think i finish with a first round for now also general comment in cases when algo is taking more then a page it would be nice to split and extract separate parts from it to make it more easier to follow, not it's can be quite time consuming to get what's going on there

Commit:
Check cursor.next result and throw if false and add missing checkAndClearBoundsFlag

Commit:
Introduced static mask variables in GenSafePointer

Commit:
Javadoc about ByteArrayPageCursor causing slower method dispatching for all PageCursors

burqen commented:
@chrisvest @MishaDemianenko Comments addressed and also cleaned up some additional stuff that we found. Go get 'em tiger 🐯

tinwelint committed:
Fixed an invalidly negated check in SeekCursor

tinwelint commented:
For the record: we've set a deadline for addressing new comments (we believe that up to this point all current comments have been addressed) to Monday 27:th. This because further development is moving quickly ahead and the gap between this PR and the further development should be limited as much as possible.
